test/unit/test_knn.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not
+# use this file except in compliance with the License. A copy of the License
+# is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import numpy as np
+import os
+import pytest
+import tempfile
+import torch as pt
+
+import sockeye
+import sockeye.constants as C
+from sockeye.generate_decoder_states import NumpyMemmapStorage, DecoderStateGenerator
+from sockeye.knn import KNNConfig, FaissIndexBuilder
+from sockeye.vocab import build_vocab
+
+# Only run certain tests in this file if faiss is installed
+try:
+    import faiss  # pylint: disable=E0401
+    faiss_installed = True
+except:
+    faiss_installed = False
+
+
+def test_numpy_memmap_storage():
+    # test open
+    with tempfile.TemporaryDirectory() as memmap_dir:
+        memmap_file = os.path.join(memmap_dir, "foo")
+        store = NumpyMemmapStorage(memmap_file, 64, np.float16)
+        store.open(64, 8)
+
+        # test add
+        ones_block_16 = np.ones((16, 64), dtype=np.float16)
+        ones_block_32 = np.ones((32, 64), dtype=np.float16)
+        zeros_block_16 = np.zeros((16, 64), dtype=np.float16)
+        zeros_block_32 = np.zeros((32, 64), dtype=np.float16)
+        # add for 0-15
+        store.add(ones_block_16)
+        assert (store.mmap[:16, :] == ones_block_16).all()
+        assert (store.mmap[16:32, :] == zeros_block_16).all()
+        assert (store.mmap[32:, :] == zeros_block_32).all()
+        # add for 16-31
+        store.add(ones_block_16)
+        assert (store.mmap[:32, :] == ones_block_32).all()
+        assert (store.mmap[32:, :] == zeros_block_32).all()
+        # add with size overflow -- this should trigger a warning without doing anything
+        store.add(np.ones((64, 64), dtype=np.float16))
+        assert (store.mmap[:32, :] == ones_block_32).all()
+        assert (store.mmap[32:, :] == zeros_block_32).all()
+
+
+def test_decoder_state_generator():
+    # utils.is_gzip_file expects byte-like object, hence supplying a byte string
+    data = 'One Ring to rule them all, One Ring to find them'
```

Comment:
Right now `data` is a `str`. Is it intended to be `bytes`?

Reply:
I forgot what caused the change exactly but seems like we can use a regular string for this test now. I should update that comment though.